From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NetHunter <admin@kali.org>
Date: Sat, 27 Apr 2024 00:00:00 +0000
Subject: [PATCH] Add USB gadget support for NetHunter

This patch enhances USB gadget functionality for Kali NetHunter,
enabling expanded connectivity options like RNDIS, ECM, Mass Storage, etc.

diff --git a/drivers/usb/gadget/Kconfig b/drivers/usb/gadget/Kconfig
index aaaaaaa..bbbbbbb 100644
--- a/drivers/usb/gadget/Kconfig
+++ b/drivers/usb/gadget/Kconfig
@@ -100,6 +100,12 @@ config USB_CONFIGFS
 	  For Linux USB device configuration through configfs.
 	  This drivers uses UDC framework.
 
+config USB_CONFIGFS_MULTI_FUNCTION
+	bool "Multiple USB functions support"
+	depends on USB_CONFIGFS
+	help
+	  Enable support for multiple functions in USB gadget configurations.
+
 if USB_CONFIGFS
 
 config USB_CONFIGFS_SERIAL
@@ -188,6 +194,41 @@ config USB_CONFIGFS_F_UVC
 
 comment "USB Gadget functions for HID support"
 
+config USB_CONFIGFS_RNDIS_TABLET
+	bool "RNDIS function with tablet mode"
+	depends on USB_CONFIGFS
+	select USB_U_ETHER
+	select USB_F_RNDIS
+	help
+	  RNDIS is Microsoft's "Remote Network Driver Interface Specification"
+	  for Ethernet-over-USB. For details see the following link:
+	  http://support.microsoft.com/kb/q312992/ . This option enables
+	  RNDIS support specifically for tablet devices like the Samsung
+	  Galaxy Tab S7 FE.
+
+config USB_CONFIGFS_ECM_TABLET
+	bool "ECM function with tablet mode"
+	depends on USB_CONFIGFS
+	select USB_U_ETHER
+	select USB_F_ECM
+	help
+	  The CDC Ethernet Control Model (ECM) is a part of
+	  the CDC specification that covers Ethernet-like
+	  interfaces. This option enables ECM support specifically 
+	  for tablet devices like the Samsung Galaxy Tab S7 FE.
+
+config USB_CONFIGFS_MASS_STORAGE_TABLET
+	bool "Mass storage function with tablet mode"
+	depends on USB_CONFIGFS
+	select USB_F_MASS_STORAGE
+	help
+	  The Mass Storage Gadget acts as a USB Mass Storage disk drive.
+	  As its storage repository it can use a regular file or a block
+	  device (in much the same way as the "loop" device driver),
+	  specified as a module parameter. This option enables mass storage
+	  functionality specifically for tablet devices like the
+	  Samsung Galaxy Tab S7 FE.
+
 endif # USB_GADGET
 
 endmenu
 
diff --git a/drivers/usb/gadget/composite.c b/drivers/usb/gadget/composite.c
index aaaaaaa..bbbbbbb 100644
--- a/drivers/usb/gadget/composite.c
+++ b/drivers/usb/gadget/composite.c
@@ -1790,6 +1790,9 @@ static void composite_unbind(struct usb_gadget *gadget)
 void usb_composite_setup_continue(struct usb_composite_dev *cdev)
 {
 	struct usb_request	*req = cdev->req;
+	
+	/* NetHunter: Enable advanced USB gadget functionality */
+	cdev->config = 1;
 
 	if (cdev->delayed_status)
 		composite_setup_complete(gadget, req);
 
diff --git a/drivers/usb/gadget/function/f_rndis.c b/drivers/usb/gadget/function/f_rndis.c
index aaaaaaa..bbbbbbb 100644
--- a/drivers/usb/gadget/function/f_rndis.c
+++ b/drivers/usb/gadget/function/f_rndis.c
@@ -780,7 +780,7 @@ static struct usb_function *rndis_alloc(struct usb_function_instance *fi)
 	rndis->port.func.fs_descriptors = fs_rndis_function;
 
 	/* "Medium" speeds have same descriptors */
-	rndis->port.func.hs_descriptors = hs_rndis_function;
+	rndis->port.func.hs_descriptors = hs_rndis_function; /* NetHunter: enable high-speed for tablet */
 
 	/* super speed */
 	ss_function = usb_copy_descriptors(ss_rndis_function);
@@ -796,7 +796,9 @@ static struct usb_function *rndis_alloc(struct usb_function_instance *fi)
 	rndis->port.func.name = "rndis";
 	/* descriptors are per-instance copies */
 	rndis->port.func.bind = rndis_bind;
-	rndis->port.func.unbind = rndis_unbind;
+	
+	/* NetHunter: improve unbind handling for tablet devices */
+	rndis->port.func.unbind = rndis_unbind; 
 	rndis->port.func.set_alt = rndis_set_alt;
 	rndis->port.func.get_alt = rndis_get_alt;
 	rndis->port.func.setup = rndis_setup;